src/osd/OSD.cc

```diff
@@ -1716,6 +1716,22 @@ void OSDService::enqueue_front(OpQueueItem&& qi)
   osd->op_shardedwq.queue_front(std::move(qi));
 }
 
+void OSDService::queue_recovery_context(
```

Comment:
When it comes to async read completion, the code is being executed by one of the `bstore_aio` threads. These guys aren't too familiar with the `OSDService` and, I bet, take many cache misses here (*might* be important in the 4 KiB reads scenario). If profiling confirms that, we could split this routine and delegate lion's share of the work to corresponding `tp_osd_tp`. What would be left to `bstore_aio` is just the injection preallocated of  `PGRecoveryContext` to `ShardedOpWQ::ShardData` using a precalculated pointer.

Reply:
CPU cyles burnt during 4 KiB randreads by `bstore_aio` threads:
```
-    5,14%     0,25%  bstore_aio  ceph-osd              [.] KernelDevice::_aio_thread
   - 4,88% KernelDevice::_aio_thread
      - 4,63% BlueStore::AioReadBatch::aio_finish
         - 1,54% OSDService::queue_recovery_context
              0,86% OSD::ShardedOpWQ::_enqueue
         - 1,43% BlueStore::PlainRegionReader::postprocess
            - 0,96% BlueStore::RegionReader::_verify_csum
                 0,76% bluestore_blob_t::verify_csum
           0,75% BlueStore::AioReadBatch::~AioReadBatc
```

Reply:
The costs of going through `shard_list` (CPU cycles):

```
OSD::ShardedOpWQ::_enqueue  /usr/local/bin/ceph-osd
  0,09 │       sar    $0x3,%rcx                                                ▒
       │     _ZNK5spg_t13hash_to_shardEj():                                    ▒
  0,18 │       div    %ecx                                                     ▒
       │     _ZN3OSD11ShardedOpWQ8_enqueueEO11OpQueueItem():                   ▒
       │       uint32_t shard_index =                                          ◆
       │         item.get_ordering_token().hash_to_shard(shard_list.size());   ▒
       │                                                                       ▒
       │       ShardData* sdata = shard_list[shard_index];                     ▒
  3,69 │       mov    %edx,%r13d                                               ▒
       │       mov    (%rsi,%r13,8),%rbx                                       ▒
       │       assert (NULL != sdata);                                         ▒
 15,94 │       test   %rbx,%rbx                                                ▒
       │     ↓ je     237                                                      ▒
  0,46 │       mov    0x8(%rbp),%eax                                           ▒
       │       unsigned priority = item.get_priority();                        ▒
       │       unsigned cost = item.get_cost();                                ▒
       │       sdata->sdata_op_ordering_lock.Lock();                           ▒
  1,66 │       lea    0x98(%rbx),%r14                                          ▒
       │       xor    %esi,%esi                                                ▒
  0,18 │       mov    0x14(%rbp),%r15d                                         ▒
  0,09 │       mov    %r14,%rdi                                                ▒
       │       mov    %eax,0x4(%rsp)                                           ▒
  0,18 │     → callq  Mutex::Lock(bool)                                        ▒
```